### PR TITLE
allow preprocessed mock data in prototype datasets tests

### PR DESCRIPTION
**Pull request by pmeier**
Currently, we enforce that the mock data functions for the prototype datasets tests generate the exact files needed for the given config. Some datasets however perform non-standard preprocessing. 

For example, the SBU archive (#5683 @lezwon) only provides a list of URLs that need to be downloaded by us before we can start yielding samples. Since we cannot perform this download in the test suite (time consuming and would fail for Meta internal systems), the test suite needs to be able to also accept the preprocessed version of the file.

This PR adds this capability. Plus it also enforces that no extra files are generated by the mock data function to avoid problems like #5549.

### CODE REVIEW

**Comment by lezwon** [comment]
@pmeier any update on when this can be merged? I'd like to continue on #5683 as soon as it's done.

**Comment by NicolasHug** [comment]
> the SBU archive (https://github.com/pytorch/vision/pull/5683 @lezwon) only provides a list of URLs that need to be downloaded by us before we can start yielding samples

Does that mean that most of the downloading happens in `__iter__()`?

**Comment by pmeier** [comment]
> Does that mean that most of the downloading happens in `__iter__()`?

Nope, the `preprocess` happens before the resource is loaded the same way as `preprocess="decompress"`. This changes nothing about the paradigm to download everything upfront.

File: test/builtin_dataset_mocks.py [hunk 1]
```diff
@@ -571,16 +590,22 @@ def generate(
...
-        for split in ("train", "val"):
-            config_name = f"{split}{year}"
+        for split_ in ("train", "val"):
+            config_name = f"{split_}{year}"
+
+            images_meta = cls._make_images_folder(root, config_name, num_samples=num_samples)
+            if split_ == split:
+                make_zip(root, f"{config_name}.zip")
+            else:
+                shutil.rmtree(root / config_name)
```
Comment (NicolasHug) [inline test/builtin_dataset_mocks.py:607]: Could you please explain what this does and why this is needed?

Comment (pmeier) [reply test/builtin_dataset_mocks.py:607]: Before this patch, the mock data generation always generated the train and the validation archive. Since only one of them is ever used by a given config, we delete the one that is not used. Instead of doing that, we could also generate the image metadata manually and only create an image folder for the correct split, rather than generate both and extract the metadata from it.

File: test/builtin_dataset_mocks.py [hunk 2]
```diff
@@ -72,6 +74,25 @@ def prepare(self, home, config):
...
+        if extra_file_names:
+            for extra in extra_file_names.copy():
+                candidate_pattern = re.compile(fr"^{extra.split('.', 1)[0]}([.]\w+)*$")
+                try:
+                    missing = next(missing for missing in missing_file_names if candidate_pattern.match(missing))
+                except StopIteration:
+                    continue
+
+                extra_file_names.remove(extra)
+                missing_file_names.remove(missing)
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py:90]: I've expanded the comment above the block. LMK if it is sufficient.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py:90]: Sorry I still don't understand the regex (do we really need a regex?), not why the raw data ends up in missing_file_names.

The problem we're having seems reasonably simple, yet the code to handle it looks fairly complex. Is there something we can do to simplify the logic here?

[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py:90]: > why the raw data ends up in missing_file_names.

The `required_file_names` include the name of the resources defined on the dataset, while the `available_file_names` just glob the directory. So if the dataset defines `OnlineResource(file_name="foo.tar")`, but we only find a folder `foo`, `foo.tar` will be in `missing_file_names` while `foo` will be in `extra_file_names`. This code block finds this correspondence and subsequently removes both from the sets.

> Is there something we can do to simplify the logic here?

Will try.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py:90]: In the example you describe, `foo.tar` was removed during the "preprocessing" stage? Or just not created at all in the mock because we don't actually need it?

[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py:90]: Let's make it more concrete: the SBU dataset provides a `SBUCaptionedPhotoDataset.tar.gz` archive, which contains a list of URLs for images and corresponding captions. Thus, before you can use the dataset, we need download all of these images. In #5683 we opted to put this functionality into the `preprocess` stage. Thus, when the dataset is loaded the first time, the archive is downloaded and the preprocessing extracts it and starts the download of all the images. Since we are not repackaging the images, we put the original data as well as the downloaded images into a `SBUCaptionedPhotoDataset` folder. With this, our automatic resource loading automatically picks the extracted folder to load the resource and thus also load the images.

> In the example you describe, `foo.tar` was removed during the "preprocessing" stage? Or just not created at all in the mock because we don't actually need it?

TL;DR: `foo.tar` is not created by the test suite.

In the test suite, we cannot perform the preprocessing, because it would require internet access and would take quite some time even if we provide fake images (this is what @lezwon had as a first draft). Thus, the easiest way to provide mock data is to provide the folder `SBUCaptionedPhotoDataset` directly rather than `SBUCaptionedPhotoDataset.tar.gz`. Thus, if the dataset loads the resource it is tricked into loading the "already preprocessed folder" rather than the raw archive.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py]: Would things be simpler if the extra download wasn't part of the `preprocess` stage?

File: test/builtin_dataset_mocks.py [hunk 3]
```diff
@@ -72,6 +71,24 @@ def prepare(self, home, config):
...
         missing_file_names = required_file_names - available_file_names
+        extra_file_names = available_file_names - required_file_names
+
+        # Some datasets need to provide already preprocessed data, for example if the preprocessing includes downloads.
+        # Such data will be included in the `extra_file_names` while the raw data will be in the `missing_file_names`.
+        # This detects these cases and removes the corresponding entries from the sets since the files are neither extra
+        # nor missing.
+        if missing_file_names:
+            for missing in missing_file_names.copy():
+                extra_candidate = missing.split(".", 1)[0]
```
Comment (NicolasHug) [inline test/builtin_dataset_mocks.py:82]: Isn't this equivalent to ?

```suggestion
                extra_candidate = missing.split(".")[0]
```

File: test/builtin_dataset_mocks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: It would make the test suite easier yes, but the resource loading more complicated. Where would you put this logic? So far this is the only dataset that has a setup like this.

File: test/builtin_dataset_mocks.py (hunk 3 above)
Comment (pmeier) [reply test/builtin_dataset_mocks.py:82]: Yes. Just a force of habit to not split more than we need. Without the `1`, split will process the whole string whereas otherwise it would stop after the first occurrence. Since we only select first part anyway, there is no need to go to the full string. Not going to fight this if you insist.

File: test/builtin_dataset_mocks.py [hunk 4]
```diff
@@ -72,6 +71,24 @@ def prepare(self, home, config):
...
+                if extra_candidate in extra_file_names:
+                    missing_file_names.remove(missing)
+                    extra_file_names.remove(extra_candidate)
```
Comment (NicolasHug) [inline test/builtin_dataset_mocks.py:85]: Is this entire thing equivalent to this?

```py
missing = set(missing.split(".")[0] for missing in missing_file_names)
not_actually_missing = missing & extra_file_names

missing -= not_actually_missing
extra_file_names -= not_actually_missing

# I guess we need to change `if missing_file_names` into `if_missing` below
```

File: test/builtin_dataset_mocks.py (hunk 3 above)
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py:82]: I had to google what this extra argument was. I'm not sure we're getting any perf boost for this, so this mostly just adds noise IMO. If we keep it, please use keyword args.

File: test/builtin_dataset_mocks.py (hunk 4 above)
Comment (pmeier) [reply test/builtin_dataset_mocks.py:85]: The logic behind it is correct, but `missing -= not_actually_missing` does not work. `not_actually_missing` contains the file names without suffixes, but `missing` has them with suffixes. Thus, we wouldn't remove anything here.

Comment (NicolasHug) [reply test/builtin_dataset_mocks.py:85]: > but missing has them with suffixes

No, `missing` is `missing_file_names` without the suffixes

File: test/builtin_dataset_mocks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py]: > Where would you put this logic?

Could we do it after the resource is loaded, withing the `make_dp()` method? 

> So far this is the only dataset that has a setup like this.

Indeed, this is why it's worth weighing whether it's worth special-casing the test suite just for this single dataset.

File: test/builtin_dataset_mocks.py (hunk 4 above)
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: True, I missed that. In that case the error message is less expressive since it will now say "`foo` is missing" although it should say `"foo.tar" is missing`.

File: test/builtin_dataset_mocks.py [hunk 5]
```diff
@@ -72,6 +71,24 @@ def prepare(self, home, config):
         available_file_names = {path.name for path in root.glob("*")}
         required_file_names = {resource.file_name for resource in self.dataset.resources(config)}
         missing_file_names = required_file_names - available_file_names
+        extra_file_names = available_file_names - required_file_names
+
+        # Some datasets need to provide already preprocessed data, for example if the preprocessing includes downloads.
+        # Such data will be included in the `extra_file_names` while the raw data will be in the `missing_file_names`.
+        # This detects these cases and removes the corresponding entries from the sets since the files are neither extra
+        # nor missing.
```
[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: This is just stripping all suffixes. So, no, this is not specific to SBU.

File: test/builtin_dataset_mocks.py (hunk 4 above)
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py]: If what we're doing is really equivalent to the snippet above, then I can't make sense out of it.

A = available
R = required
M = missing = R - A
E = extra = A - R

So the intersection between M and E is null, right?

Comment (pmeier) [reply test/builtin_dataset_mocks.py]: Correct. And that is why we perform this matching. For example

```py
A = {"foo", "bar"}
R = {"foo.zip", "spam.zip"}
M = R - A  # {"foo.zip", "spam.zip"}
E = A - R  # {"foo", "bar"}

missing = next(iter(M))  # foo.zip
extra_candidate = missing_candidate.split(".")[0]  # foo
extra_candidate in E  # True
M.remove(missing)  # {"spam.zip"}
E.remove(extra_candidate)  # {"bar"}
```

File: test/builtin_dataset_mocks.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py]: This still assumes that the `SBUCaptionedPhotoDataset.tar.gz` contains the link to download resources that will be put into the  `SBUCaptionedPhotoDataset` folder, with a matching name, right?

[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: The folder ` SBUCaptionedPhotoDataset` is a superset of `SBUCaptionedPhotoDataset.tar.gz`. It contains all files the archive contains plus the downloaded images. So, yes, the URL list is also contained in the folder.

File: test/builtin_dataset_mocks.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: Not sure what you mean here? Do you want a general solution on dataset side? Because for the test suite this PR is a general solution for all datasets that have a similar setup. Nothing is special about SBU here.

[the remark this answers is not in the document]
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py]: > this PR is a general solution for all datasets that have a similar setup

How many datasets have a similar setup? From above it seems like this is the only one. Hence I'm wondering whether having this custom logic, on which you and I have spent a few hours, already is really necessary.

[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: Prototype for loading inside `_make_dp` is in lezwon/vision#1.

[the remark this answers is not in the document]
Comment (pmeier) [reply test/builtin_dataset_mocks.py]: > How many datasets have a similar setup? From above it seems like this is the only one.

Nope, at least Kinetics also has a similar setup:

https://github.com/pytorch/vision/blob/095cabb76cdcd4763bad629481d189b91e3df42c/torchvision/datasets/kinetics.py#L181-L190

> Hence I'm wondering whether having this custom logic, on which you and I have spent a few hours, already is really necessary.

What is the alternative? Dropping support for the dataset? We might be able to argue that SBU is not important enough to "break our rules", but Kinetics certainly is.
